Validate schema of embedded examples in a spec

NobodysNightmare opened the pull request:
Our schemas often contain embedded examples that show what a representation will look like.
These specs ensure that those embedded examples conform to the schema that they are an example for.

Some collection examples contained abbreviated examples for what they look like. To support these, the previously added _hint key was replaced with an explicit _abbreviated key and examples that include this key are skipped.

# Ticket
https://community.openproject.org/wp/69707

NobodysNightmare committed:
Validate schema of embedded examples in a spec

Our schemas often contain embedded examples that show what a representation
will look like.
These specs ensure that those embedded examples conform to the schema
that they are an example for.

Some collection examples contained abbreviated examples for what they look like.
To support these, the previously added _hint key was replaced with an explicit
_abbreviated key and examples that include this key are skipped.

NobodysNightmare commented on spec/docs/embedded_schema_examples_spec.rb:
```diff
@@ -0,0 +1,55 @@
+RSpec.describe "Examples embedded in APIv3 schemas" do # rubocop:disable RSpec/DescribeClass
+  Dir[
+    Rails.root.join("docs/api/apiv3/components/schemas/*").to_s
+  ].map { |f| File.basename(f).split(".", 2).first }.grep(/model$/).each do |schema_name|
```
:yellow_circle: There should be a self-test that ensures the auto-discovery is actually finding schemas. Otherwise restructuring the docs files could break this test unnoticed.

Kharonus commented on spec/docs/embedded_schema_examples_spec.rb:
```diff
@@ -0,0 +1,63 @@
+  schema_names.each do |schema_name|
+    it "has no schema errors in #{schema_name}" do
+      schema = JsonSchemaLoader.new.load(schema_name)
+      example = schema[:example]
```
🟡 If I read this correctly, this only tests the examples defined below the `example` keyword.

OpenAPI - and also our spec - does also contain the `examples` key, where you have a map of examples. 

❓ Will this also resolves `$ref` if this is given in the example?

Kharonus replied:
To be clear: I'd still approve this, as this is already a great improvement. See it more like a next step for improvement.

NobodysNightmare replied:
Yeah, it's a good point. I didn't think that multiple examples would've been possible, in this context. It's important to note that these are embedded in (i.e. part of) the schema. The normal "OpenAPI case" is one that I am covering in a separate PR.

Still you are right, and I see us using `examples` in the schema as well. Interesting things I am noticing:

* I can find only four cases, where we have `examples` directly embedded in the schema
* In all those cases they are an array of examples, not a map
* In all those cases, we only have exactly one example

Right now I am considering to just change those cases to use the `example` syntax as well.

NobodysNightmare replied:
> ❓ Will this also resolves $ref if this is given in the example?

It would indeed. I'd need to check the JSON schema spec on this, no clue whether that's intended or something we should avoid. Right now it doesn't seem to matter.

Kharonus commented on docs/api/apiv3/components/schemas/work_package_model.yml:
```diff
@@ -54,14 +54,14 @@ properties:
     description: Date on which a milestone is achieved
   derivedStartDate:
-    type: string
+    type: [string, null]
```
🟡 what about the discussion with block-style vs. flow-style and the indentation of arrays?

Should we ignore that or do we accept a common pattern?

NobodysNightmare replied:
Thanks for checking this. I probably just forgot it. Parts of this have been written/corrected before our initial discussion. I will fix it.

Kharonus commented on docs/api/apiv3/components/schemas/work_package_model.yml:
```diff
@@ -54,14 +54,14 @@ properties:
     readOnly: true
   derivedDueDate:
-    type: string
+    type: [string, null]
```
🔴 At least `null` must be defined as a string, not a literal.

NobodysNightmare replied:
fixed

Kharonus commented on docs/api/apiv3/components/schemas/query_filter_instance_schema_model.yml:
```diff
@@ -56,19 +53,7 @@ example:
-    hasDefault: false
-    writable: true
-    _links: {}
+  name: "A name"
```
🟢 Haha, nice catch. 😂

Kharonus reviewed (changes requested):
nice idea of automatically testing examples!

NobodysNightmare committed:
Change a few occurences of examples to example

All of them only contained a single example anyways. This
change immediately enabled validation of the example against
the schema in specs, which uncovered a wrong schema.

Examples and the implementation of the TimeEntryRepresenter
agreed that the comment was a formattable, but the schema claimed
it would be a string.

NobodysNightmare committed:
Fix formatting of type

Kharonus reviewed (approved):
LGTM
